gmail/stream_implementer.bal

```diff
@@ -0,0 +1,317 @@
+        //Get json thread list reponse. If unsuccessful throws and returns error.
+        json jsonListThreadResponse = check handleResponse(httpResponse);
+        ThreadListPage|error response = jsonListThreadResponse.cloneWithType(ThreadListPage);
+
+        if (response is ThreadListPage) {
+            self.pageToken = response?.nextPageToken;
+            var threads = response?.threads;
+            if (threads is MailThread[]) {
+                return threads;
+            }
+            return [];
+        } else {
+            return error(ERR_THREAD_LIST, response);
+        }
+    }
+}
+
+class DraftStream {
+    private Draft[] currentEntries = [];
+    private int index = 0;
+    private final http:Client httpClient;
+    private final string userEmailId;
+    private final MsgSearchFilter? filter;
+    private string? pageToken = ();
+
+
+    isolated function init(http:Client httpClient, string userId, MsgSearchFilter? filter) returns error? {
+        self.httpClient = httpClient;
+        self.filter = filter;
+        if (self.filter?.pageToken is string) {
+            self.pageToken = <string>self.filter?.pageToken;
+        }
+        self.userEmailId = userId;
+        self.currentEntries = check self.fetchDrafts();
+    }
+
+    public isolated function next() returns @tainted record {| Draft value; |}|error? {
+        if (self.index < self.currentEntries.length()) {
+            record {| Draft value; |} draft = {value: self.currentEntries[self.index]};
+            self.index += 1;
+            return draft;
+        }
+        if (self.pageToken is string) {
+            self.index = 0;
+            self.currentEntries = check self.fetchDrafts();
+            record {| Draft value; |} draft = {value: self.currentEntries[self.index]};
+            self.index += 1;
+            return draft;
+        }
+    }
+
+    isolated function fetchDrafts() returns @tainted Draft[]|error {
+
+        string getListDraftsPath = USER_RESOURCE + self.userEmailId + DRAFT_RESOURCE;
+        if (self.filter is DraftSearchFilter) {
+            string uriParams = "";
+            //The default value for include spam trash query parameter of the api call is false
+            uriParams = self.filter?.includeSpamTrash is boolean ? check appendEncodedURIParameter(uriParams, INCLUDE_SPAMTRASH,
+                string `${<boolean>self.filter?.includeSpamTrash}`) : uriParams;
+            uriParams = self.filter?.maxResults is int ? check appendEncodedURIParameter(uriParams, MAX_RESULTS, 
+                self.filter?.maxResults.toString()) : uriParams;
+            uriParams = self.pageToken is string ? check appendEncodedURIParameter(uriParams, PAGE_TOKEN, 
+                <string>self.pageToken) :uriParams;
+            uriParams = self.filter?.q is string ? check appendEncodedURIParameter(uriParams, QUERY, <string>self.filter?.q) : 
+                uriParams;
+            getListDraftsPath += <@untainted>uriParams;
```

Comment:
Shall we create function to handle this as it is used in some other places as well? line 145